hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/prometheus/RestTemplateParallelPrometheusClient.java

```diff
@@ -0,0 +1,114 @@
+        this.parallelFetchingTimeout = parallelFetchingTimeoutMillis;
+        this.meterRegistry = meterRegistry;
+    }
+
+    @Override
+    public MonitoringMetricsContainer readMetrics(List<String> queries) {
+        return fetchInParallelFromPrometheus(queries);
+    }
+
+    private MonitoringMetricsContainer fetchInParallelFromPrometheus(List<String> queries) {
+        CompletableFuture<Map<String, MetricDecimalValue>> aggregatedFuture = getAggregatedCompletableFuture(queries);
+
+        try {
+            Map<String, MetricDecimalValue> metrics = aggregatedFuture.get(parallelFetchingTimeout.toMillis(), TimeUnit.MILLISECONDS);
+            return MonitoringMetricsContainer.initialized(metrics);
+        } catch (InterruptedException e) {
+            // possibly let know the caller that the thread was interrupted
+            Thread.currentThread().interrupt();
+            logger.warn("Prometheus fetching thread was interrupted...", e);
+            return MonitoringMetricsContainer.unavailable();
+        } catch (Exception ex) {
+            logger.warn("Unexpected exception during fetching metrics from prometheus...", ex);
+            return MonitoringMetricsContainer.unavailable();
+        }
+    }
+
+    private CompletableFuture<Map<String, MetricDecimalValue>> getAggregatedCompletableFuture(List<String> queries) {
+        // has to be collected to run in parallel
+        List<CompletableFuture<Pair<String, MetricDecimalValue>>> futures = queries.stream()
+                .map(this::readSingleMetric)
+                .toList();
+
+        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+                .thenApply(
+                        v -> futures.stream().map(CompletableFuture::join)
+                                .collect(Collectors.toMap(Pair::getKey, Pair::getValue))
+                );
+    }
+
+    private CompletableFuture<Pair<String, MetricDecimalValue>> readSingleMetric(String query) {
+        return CompletableFuture.supplyAsync(() -> queryPrometheus(query), executorService);
+    }
+
+    private Pair<String, MetricDecimalValue> queryPrometheus(String query) {
+        try {
+            URI queryUri = URI.create(prometheusUri.toString() + "/api/v1/query?query=" + encode(query, UTF_8));
+            PrometheusResponse response = restTemplate.exchange(queryUri,
+                    HttpMethod.GET, HttpEntity.EMPTY, PrometheusResponse.class).getBody();
+
+            Preconditions.checkNotNull(response, "Prometheus response is null");
+            Preconditions.checkState(response.isSuccess(), "Prometheus response does not contain valid data");
+
+            MetricDecimalValue result = parseResponse(response);
+            meterRegistry.counter("read-metric-from-prometheus.success").increment();
+            return Pair.of(query, result);
+        } catch (Exception exception) {
+            logger.warn("Unable to read from Prometheus...", exception);
```

Comment:
we should log `query` here so we would be able to tell which query is responsible